File: digidis-go/day10/day10.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func main() {
	part1()
	part2()
}

func part1() {
	var lengths = strings.Split("183,0,31,146,254,240,223,150,2,206,161,1,255,232,199,88", ",")

	list := make([]int, 256)
	for i := 0; i < 256; i++ {
		list[i] = i
	}
	var pos, skip int

	for _, sl := range lengths {
		l, _ := strconv.Atoi(sl)

		reverse := make([]int, l)
		for k := 0; k < l; k++ {
			reverse[k] = list[(pos+k)%256]
		}
		for k := 0; k < l; k++ {
			list[(pos+k)%256] = reverse[l-k-1]
		}
		pos = (pos + l + skip) % 256
		skip++
	}

	fmt.Printf("Result 1 %v\n", list[0]*list[1])
}

func part2() {

	var input = "183,0,31,146,254,240,223,150,2,206,161,1,255,232,199,88"
	var lengths []int
	for l := range input {
		lengths = append(lengths, int(input[l]))
	}
	lengths = append(lengths, []int{17, 31, 73, 47, 23}...)

	list := make([]int, 256)

	for i := 0; i < 256; i++ {
		list[i] = i
	}
	var pos, skip int

	for round := 0; round < 64; round++ {
		for _, l := range lengths {

			reverse := make([]int, l)
			for k := 0; k < l; k++ {
				reverse[k] = list[(pos+k)%256]
			}
			for k := 0; k < l; k++ {
				list[(pos+k)%256] = reverse[l-k-1]
			}
			pos = (pos + l + skip) % 256
			skip++
		}
	}

	var hash [16]byte

	for u := 0; u < 16; u++ {
		h := byte(list[u*16+0])
		for v := 1; v < 16; v++ {
			h = h ^ byte(list[u*16+v])
		}
		hash[u] = h
	}

	fmt.Printf("Result 2 %x\n", hash)
}

File: meldanya-python3/day10/day10.py
def part1(list_, lengths, times=1):
    pos, skip = 0, 0
    for len_ in lengths*times:
        stop = pos + len_
        wrap = 0
        if stop > len(list_):
            wrap = stop % len(list_)
            stop = len(list_)

        li = list_[pos:stop]
        if wrap > 0:
            li += list_[:wrap]
        li = list(reversed(li))

        list_[pos:stop] = li[:stop-pos]
        if wrap > 0:
            list_[:wrap] = li[stop-pos:]

        pos = (pos+len_+skip) % len(list_)
        skip += 1
    return list_


def _xor(nums):
    res = 0
    for n in nums:
        res ^= n
    return res


def part2(list_, lengths):
    lengths = [ord(c) for c in lengths]
    lengths.extend([17, 31, 73, 47, 23])
    res = part1(list_[:], lengths, times=64)
    dense = [0]*16
    for i in range(16):
        dense[i] = _xor(res[i*16:i*16+16])
    return ''.join(['%02x' % d for d in dense])


def main():
    line = input()
    list_ = list(range(256))

    lengths = [int(l) for l in line.split(',')] if line else []
    p1 = part1(list_[:], lengths)
    print(p1[0] * p1[1])
    print(part2(list_[:], line))


if __name__ == '__main__':
    main()
